src/samples/mqtt/bluemix.c

```diff
@@ -0,0 +1,355 @@
+
+    return false;
+
+error:
+    SOL_WRN("Unable to initialize MQTT");
+    sol_quit();
+
+    return false;
+}
+
+/**
+ * Authenticate to HTTP Server
+ */
+
+static bool
+json_token_to_string(struct sol_json_token *token, char **out)
+{
+    if (sol_json_token_get_type(token) != SOL_JSON_TYPE_STRING)
+        return false;
+    free(*out);
+    *out = strndup(token->start + 1, token->end - token->start - 2);
+    return !!*out;
+}
+
+static bool
+parse_auth_json(struct bluemix_data *data, const char *json, size_t size)
+{
+    struct sol_json_scanner scanner;
+    struct sol_json_token token, key, value;
+    enum sol_json_loop_reason reason;
+
+    sol_json_scanner_init(&scanner, json, size);
+
+    SOL_JSON_SCANNER_OBJECT_LOOP (&scanner, &token, &key, &value, reason) {
+        if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "clientid")) {
+            if (!json_token_to_string(&value, &data->client_id))
+                goto error;
+        } else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "authtoken")) {
+            if (!json_token_to_string(&value, &data->auth_token))
+                goto error;
+        } else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "orgid")) {
+            if (!json_token_to_string(&value, &data->org_id))
+                goto error;
+        }
+    }
+
+    return true;
+
+error:
+    SOL_WRN("Error parsing auth json");
+    return false;
+}
+
+static bool get_user_token(void *vdata);
+
+static void
+request_callback(void *vdata, const struct sol_http_client_connection *connection,
+    struct sol_http_response *response)
+{
+    struct bluemix_data *data = (struct bluemix_data *)vdata;
+
+    if (!response->content.used || response->response_code != SOL_HTTP_STATUS_OK) {
+        SOL_DBG("Message: %.*s", (int)response->content.used, (char *)response->content.data);
+        goto error;
+    }
+
+    if (!parse_auth_json(data, response->content.data, response->content.used))
+        goto error;
+
+    SOL_INF("Connected: Starting sensor communication");
+
+    sol_timeout_add(1000, mqtt_init, data);
+
+    return;
+
+error:
+    SOL_WRN("Unable to get client id and auth code. Retrying...");
+    if (!data->connect_timeout) {
+        data->connect_timeout = sol_timeout_add(1000, get_user_token, data);
+
+        if (!data->connect_timeout) {
+            SOL_WRN("Retry failed.");
+            sol_quit();
+        }
+    }
+}
+
+static bool
+get_user_token(void *vdata)
+{
+    struct sol_http_param params;
+    struct sol_http_client_connection *connection;
+    struct sol_buffer buffer = SOL_BUFFER_INIT_EMPTY;
+    struct bluemix_data *data = (struct bluemix_data *)vdata;
+    int r;
+
+    SOL_NULL_CHECK_GOTO(data, error);
+    SOL_NULL_CHECK_GOTO(data->user, error);
+    SOL_NULL_CHECK_GOTO(data->pin, error);
+    SOL_NULL_CHECK_GOTO(data->url, error);
+
+    data->connect_timeout = NULL;
+
+    r = sol_buffer_append_printf(&buffer, "{\"email\":\"%s\", \"pin\":\"%s\"}", data->user, data->pin);
```

Comment:
similar to above, sync with @ceolin to use the string serialize function as the proper way to do this is to escape the string.